llap-server/src/java/org/apache/hadoop/hive/llap/io/encoded/LowLevelDataReader.java

```diff
@@ -0,0 +1,135 @@
+import org.apache.orc.impl.OrcCodecPool;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.function.Supplier;
+
+public class LowLevelDataReader implements AutoCloseable {
+
+  private Path path;
+  private Object fileKey;
+  private Configuration daemonConf;
+  private DataCache cache;
+  private FileMetadataCache metadataCache;
+  private CacheTag cacheTag;
+  private FixedSizedObjectPool<IoTrace> tracePool;
+
+  private Supplier<FileSystem> fsSupplier;
+
+  private Reader orcReader;
+  private LlapDataReader rawDataReader;
+  private EncodedReader encodedReader;
+
+
+  public LowLevelDataReader(Path path, Object fileKey, Configuration daemonConf, DataCache cache,
+      FileMetadataCache metadataCache, CacheTag cacheTag, FixedSizedObjectPool<IoTrace> tracePool) {
+    this.path = path;
+    this.fileKey = fileKey;
+    this.daemonConf = daemonConf;
+    this.cache = cache;
+    this.metadataCache = metadataCache;
+    this.cacheTag = cacheTag;
+    this.tracePool = tracePool;
+  }
+
+  public void init() throws IOException {
+    this.fsSupplier = getFsSupplier(path, daemonConf);
+    Object fileKey = HdfsUtils.getFileId(fsSupplier.get(), path,
+          HiveConf.getBoolVar(daemonConf, HiveConf.ConfVars.LLAP_CACHE_ALLOW_SYNTHETIC_FILEID),
+          HiveConf.getBoolVar(daemonConf, HiveConf.ConfVars.LLAP_CACHE_DEFAULT_FS_FILE_ID),
+          !HiveConf.getBoolVar(daemonConf, HiveConf.ConfVars.LLAP_IO_USE_FILEID_PATH));
+    if(!fileKey.equals(this.fileKey)) {
+      throw new IOException("File key mismatch.");
+    }
+    OrcFile.ReaderOptions opts = EncodedOrcFile.readerOptions(daemonConf).filesystem(fsSupplier);
+    orcReader = EncodedOrcFile.createReader(path, opts);
+  }
+
+  private static Supplier<FileSystem> getFsSupplier(final Path path, final Configuration conf) {
+    return () -> {
+      try {
+        return path.getFileSystem(conf);
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    };
+  }
+
+  public void read(DiskRangeList ranges) throws IOException {
+    boolean useZeroCopy = (daemonConf != null) && OrcConf.USE_ZEROCOPY.getBoolean(daemonConf);
+    InStream.StreamOptions options = InStream.options()
+        .withCodec(OrcCodecPool.getCodec(orcReader.getCompressionKind()))
+        .withBufferSize(orcReader.getCompressionSize());
+
+    rawDataReader = LlapRecordReaderUtils.createDefaultLlapDataReader(
+        DataReaderProperties.builder()
+            .withFileSystemSupplier(fsSupplier).withPath(path)
+            .withCompression(options)
+            .withZeroCopy(useZeroCopy)
+            .build());
+    rawDataReader.open();
+
+    IoTrace ioTrace = tracePool.take();
+
+    try {
+      encodedReader = orcReader.encodedReader(fileKey, cache, rawDataReader, null, ioTrace, false, cacheTag, false);
+      encodedReader.readDataRanges(ranges);
+    } finally {
+      tracePool.offer(ioTrace);
+    }
+  }
+
+  public void readFooter() {
```

Comment:
I think we could return the tailBuffers here, after all this is a reader class :) - caller will decide what to do with the result later